File: frontend/lib/models/document.dart
enum DocumentType { pdf, image, document }

enum DocumentStatus { uploaded, processing, completed, failed }

class Document {
  final int id;
  final int ownerId;
  final String fileName;
  final String filePath;
  final DocumentType fileType;
  final int fileSize;
  final DocumentStatus status;
  final String? extractionMethod;
  final String? extractedText;
  final String? aiSummary;
  final String? aiModelUsed;
  final int processingProgress;
  final DateTime createdAt;
  final DateTime? processingStartedAt;
  final DateTime? processingCompletedAt;

  Document({
    required this.id,
    required this.ownerId,
    required this.fileName,
    required this.filePath,
    required this.fileType,
    required this.fileSize,
    required this.status,
    this.extractionMethod,
    this.extractedText,
    this.aiSummary,
    this.aiModelUsed,
    required this.processingProgress,
    required this.createdAt,
    this.processingStartedAt,
    this.processingCompletedAt,
  });

  factory Document.fromJson(Map<String, dynamic> json) {
    return Document(
      id: json['id'] as int,
      ownerId: json['owner_id'] as int,
      fileName: json['file_name'] as String,
      filePath: json['file_path'] as String,
      fileType: DocumentType.values.firstWhere(
        (e) => e.name == (json['file_type'] as String?)?.toLowerCase(),
        orElse: () => DocumentType.document,
      ),
      fileSize: json['file_size'] as int? ?? 0,
      status: DocumentStatus.values.firstWhere(
        (e) => e.name == (json['status'] as String?)?.toLowerCase(),
        orElse: () => DocumentStatus.uploaded,
      ),
      extractionMethod: json['extraction_method'] as String?,
      extractedText: json['extracted_text'] as String?,
      aiSummary: json['ai_summary'] as String?,
      aiModelUsed: json['ai_model_used'] as String?,
      processingProgress: json['processing_progress'] as int? ?? 0,
      createdAt: DateTime.parse(json['created_at'] as String),
      processingStartedAt: json['processing_started_at'] != null
          ? DateTime.parse(json['processing_started_at'] as String)
          : null,
      processingCompletedAt: json['processing_completed_at'] != null
          ? DateTime.parse(json['processing_completed_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'owner_id': ownerId,
      'file_name': fileName,
      'file_path': filePath,
      'file_type': fileType.name,
      'file_size': fileSize,
      'status': status.name,
      'extraction_method': extractionMethod,
      'extracted_text': extractedText,
      'ai_summary': aiSummary,
      'ai_model_used': aiModelUsed,
      'processing_progress': processingProgress,
      'created_at': createdAt.toIso8601String(),
      'processing_started_at': processingStartedAt?.toIso8601String(),
      'processing_completed_at': processingCompletedAt?.toIso8601String(),
    };
  }
}

File: backend/app/models/document.py
"""
Document model
"""

from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Boolean, Enum as SQLEnum
from sqlalchemy.orm import relationship
from datetime import datetime
from enum import Enum
from app.db.base import Base


class DocumentStatus(str, Enum):
    """Status of document processing"""
    UPLOADED = "uploaded"          # Just uploaded
    PROCESSING = "processing"      # Being processed by AI
    COMPLETED = "completed"        # Processing complete
    FAILED = "failed"              # Processing failed


class DocumentType(str, Enum):
    """Type of document"""
    PDF = "pdf"
    IMAGE = "image"
    DOCUMENT = "document"  # docx, etc


class Document(Base):
    """Document model - represents uploaded files for extraction"""
    __tablename__ = "documents"

    id = Column(Integer, primary_key=True, index=True)
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    
    # File info
    file_name = Column(String(255), nullable=False)
    file_path = Column(String(512), nullable=False, unique=True)
    file_type = Column(SQLEnum(DocumentType), nullable=False)
    file_size = Column(Integer)  # in bytes
    
    # Processing
    status = Column(SQLEnum(DocumentStatus), default=DocumentStatus.UPLOADED, nullable=False, index=True)
    processing_progress = Column(Integer, default=0)  # 0-100%
    
    # Extracted content
    extracted_text = Column(Text, nullable=True)
    extraction_method = Column(String(50), nullable=True)  # ocr, pdf_text, etc
    
    # AI processing
    ai_summary = Column(Text, nullable=True)
    ai_model_used = Column(String(100), nullable=True)  # gpt-4, etc
    
    # Metadata
    original_upload_time = Column(DateTime, default=datetime.utcnow, nullable=False)
    processing_started_at = Column(DateTime, nullable=True)
    processing_completed_at = Column(DateTime, nullable=True)
    
    # Error handling
    error_message = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    owner = relationship("User", back_populates="documents")
    schedule_items = relationship("ScheduleItem", back_populates="document")

    def __repr__(self):
        return f"<Document(id={self.id}, file_name={self.file_name}, status={self.status})>"
